libs/base/Data/Bits.idr

```diff
@@ -0,0 +1,192 @@
+  bit         = (1 `shiftL`)
+  zeroBits    = 0
+  testBit x i = (x .&. bit i) /= 0
+  shiftR x    = prim__shr_Bits16 x . fromInteger . cast . fst
+  shiftL x    = prim__shl_Bits16 x . fromInteger . cast . fst
+
+  popCount x0 =
+    -- see https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
+    let x1 = (x0 .&. 0x5555) + ((x0 `shiftR` fromNat 1) .&. 0x5555)
+        x2 = (x1 .&. 0x3333) + ((x1 `shiftR` fromNat 2) .&. 0x3333)
+        x3 = ((x2 + (x2 `shiftR` fromNat 4)) .&. 0x0F0F)
+        x4 = (x3 * 0x0101) `shiftR` fromNat 8
+     in fromInteger $ cast x4
+
+public export %inline
+Bits Bits32 where
+  Index       = Subset Nat (`LT` 32)
+  (.&.)       = prim__and_Bits32
+  (.|.)       = prim__or_Bits32
+  xor         = prim__xor_Bits32
+  bit         = (1 `shiftL`)
+  zeroBits    = 0
+  testBit x i = (x .&. bit i) /= 0
+  shiftR x    = prim__shr_Bits32 x . fromInteger . cast . fst
+  shiftL x    = prim__shl_Bits32 x . fromInteger . cast . fst
+
+  popCount x0 =
+    -- see https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
+    let x1 = (x0 .&. 0x55555555) + ((x0 `shiftR` fromNat 1) .&. 0x55555555)
+        x2 = (x1 .&. 0x33333333) + ((x1 `shiftR` fromNat 2) .&. 0x33333333)
+        x3 = ((x2 + (x2 `shiftR` fromNat 4)) .&. 0x0F0F0F0F)
+        x4 = (x3 * 0x01010101) `shiftR` fromNat 24
+     in fromInteger $ cast x4
+
+public export %inline
+Bits Int where
+  Index       = Subset Nat (`LT` 64)
+  (.&.)       = prim__and_Int
+  (.|.)       = prim__or_Int
+  xor         = prim__xor_Int
+  bit         = (1 `shiftL`)
+  zeroBits    = 0
+  testBit x i = (x .&. bit i) /= 0
+  shiftR x    = prim__shr_Int x . cast . fst
+  shiftL x    = prim__shl_Int x . cast . fst
+
+  popCount x =
+    -- see https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
+    -- We have to treat negative numbers separately in order to
+    -- prevent overflows in the first addition.
+    -- The top bit is therefore cleared and 1 is added in the end
+    -- in case of a negative number
+    let x0 = x `clearBit` fromNat 63
+        x1 = (x0 .&. 0x5555555555555555)
+           + ((x0 `shiftR` fromNat 1) .&. 0x5555555555555555)
+        x2 = (x1 .&. 0x3333333333333333)
+           + ((x1 `shiftR` fromNat 2) .&. 0x3333333333333333)
+        x3 = ((x2 + (x2 `shiftR` fromNat 4)) .&. 0x0F0F0F0F0F0F0F0F)
+        x4 = (x3 * 0x0101010101010101) `shiftR` fromNat 56
+        x5 = if x < 0 then x4 + 1 else x4
+     in fromInteger $ cast x5
+
+--------------------------------------------------------------------------------
+--          FiniteBits
+--------------------------------------------------------------------------------
+
+public export
+interface Bits a => FiniteBits a where
+  ||| Return the number of bits in values of type `t`.
+  bitSize : Nat
+
+  ||| Returns the bitwise complement of a value.
+  complement : a -> a
+
```

Comment:
A tempting thing to ask is why not to have a function connecting a `bitSize` or `FiniteBits` with `Index` type in `Bits` like
```idris
0 bitToIndex : Fin bitSize -> Index
```
I can imagine why they might not be isomorphic, but the function working at least one way should exist.

Reply:
Could you please elaborate on this? Do you see a use case for `bitToIndex` other than as a proof that `bitSize` and `Index` are properly correlated? Also, why might they not be isomorphic?

Reply:
> Do you see a use case for bitToIndex other than as a proof that bitSize and Index are properly correlated?

Well, no, what I originally suggested is purely a proof of proper correlation and that's why it has `0` quantity.

However, it can be useful for generic programming, i.e. for code working with any type satisfying `FiniteBits`. Now you cannot use any shifting and bit setting functions unless you know exactly the type. Having a `bitToIndex` function you can have generic functions (even if you'll need some requirement on `bitSize` value in its signature). But then, `bitsToIndex` shouldn't have zero quantity.

> Also, why might they not be isomorphic?

I can imagine a proper instance for finite-bits-type (like, e.g. `Bits32`) with `Index` being `Integer`. In this case there is no proper function of type `Index -> Fin bitSize`. This is very artificial example and maybe it's better to require an isomorphism between `Fin bitSize` and `Index` -- I'm just not sure about this.

When requiring an isomorphism, probably it's better to have a special data structure coding it (i.e. having functions in both directions and showing extensional equality of applications of these functions to each other with the identity function).

Reply:
Thanks, buzden, for your kind explanations.

Since this PR is already quite big enough, I'd rather not blow it up any more by adding a new data type/interface for `Isomorphism`. But I'll give `bitsToIndex` a go, as it makes sense to correlate the two types.

Reply:
`Fin` is not efficient at the moment so you may want to have

```idris
bitToIndex : Subset Nat (`LT` bitSize) -> Index
```

instead.

Reply:
Changed `bitToIndex` as suggested by gallais.